content/en/blog/2021-12-15-jit-costing.md

```diff
@@ -0,0 +1,225 @@
+---
+title: "Towards Ergo v5.0: JIT costing"
+date: 2021-12-15T16:02:39.774Z
+author: Ergo Foundation
+authorPhoto: /img/uploads/1762.png
+blogPhoto: 
+tags:
+---
+<!--StartFragment-->
+
+## Introduction
+
+In Ergo blockchain all crypto assets are stored in boxes protected by guarding conditions
+(aka contracts). Every unspent box is protected by a script, which is stored in
+the box itself as serialized [ErgoTree](https://ergoplatform.org/docs/ErgoTree.pdf). The
+script should be evaluated to `true` value as part of transaction validation to verify
+box's spending conditions.
+
+Script verification takes some time and system resources on every network node and thus
+both script's execution time and the resources it needs must be put under control. This is
+known as cost estimation (or "costing").
+
+Costing is necessary to prevent long delays in verification of new blocks by network
+nodes. Both over-sized and over-expensive scripts may be exploited to attack the network
+by exhausting resources of verifiers and slowing down network transaction throughput.
+
+There are two general approaches to do cost control:
+1) Ahead-Of-Time Costing (AOTC) - estimation of costs _before_ script execution, so that
+over-expensive script is not executed at all.
+2) Just-In-Time Costing (JITC) - accumulation of costs _during_ (i.e. in the process of)
+   the actual execution of a script.
+
+Ergo v5.0 will take advantage of both approaches so that verification of Ergo contracts is
+done with a hybrid JIT+AOT cost control.
```

Comment:
What does this hybrid approach look like? ie, 'This hybrid approach performs JITC first and then Ahead-Of-Time and before actual execution of sigma protocol

Reply:
Yes, something like this. This is explained in the New Costing in v5.0 section below, but probably make sense to give a hint here as well.
See https://github.com/ergoplatform/website/pull/787/files#diff-6db8af54183bd8173533453ce8974cdd702340c16c44cc8830d7950b69b54e84R171